src/env/hedged/mod.rs

```diff
@@ -0,0 +1,671 @@
+                    self.seqno2.clone(),
+                )
+            }),
+            res2 = f2 => res2.unwrap().map(|res| {
+                let (fd, is_for_rewrite) = resolve(res);
+                HedgedHandle::new(
+                    self.base.clone(),
+                    is_for_rewrite,
+                    self.sender.clone(),
+                    FutureHandle::new(f1, task1),
+                    FutureHandle::new_owned(fd) ,
+                    self.seqno1.clone(),
+                    self.seqno2.clone(),
+                )
+            }),
+        }
+    }
+
+    async fn wait(&self, task1: Task, task2: Task) -> IoResult<()> {
+        let (cb1, mut f1) = paired_future_callback();
+        let (cb2, mut f2) = paired_future_callback();
+        self.sender.send(task1, task2, cb1, cb2);
+
+        select! {
+            res1 = f1 => res1.unwrap().map(|_| ()),
+            res2 = f2 => res2.unwrap().map(|_| ()),
+        }
+    }
+}
+
+impl Drop for HedgedFileSystem {
+    fn drop(&mut self) {
+        block_on(self.wait(Task::Stop, Task::Stop)).unwrap();
+
+        let t1 = self.thread1.take().unwrap();
+        let t2 = self.thread2.take().unwrap();
+        let mut times = 0;
+        loop {
+            if t1.is_finished() && t2.is_finished() {
+                t1.join().unwrap();
+                t2.join().unwrap();
+                break;
+            }
+            times += 1;
+            if times > 100 {
+                // wait 1s
+                // one disk may be blocked for a long time,
+                // to avoid block shutdown process for a long time, do not join the threads
+                // here, only need at least to ensure one thread is exited
+                if t1.is_finished() || t2.is_finished() {
+                    if t1.is_finished() {
+                        t1.join().unwrap();
+                    } else {
+                        t2.join().unwrap();
+                    }
+                    break;
+                }
+            }
+            std::thread::sleep(std::time::Duration::from_millis(10));
+        }
+    }
+}
+
+impl FileSystem for HedgedFileSystem {
+    type Handle = HedgedHandle;
+    type Reader = HedgedReader;
+    type Writer = HedgedWriter;
+
+    fn bootstrap(&self) -> IoResult<()> {
+        // catch up diff
+        if !self.path1.exists() {
+            info!("Create raft log directory: {}", self.path1.display());
+            fs::create_dir(&self.path1).unwrap();
+        }
+        if !self.path2.exists() {
+            info!("Create raft log directory: {}", self.path2.display());
+            fs::create_dir(&self.path2).unwrap();
+        }
+        let files1 = recover::get_files(&self.path1)?;
+        let files2 = recover::get_files(&self.path2)?;
+
+        let count1 = recover::get_latest_valid_seq(&self.base, &files1)?;
+        let count2 = recover::get_latest_valid_seq(&self.base, &files2)?;
+
+        match count1.cmp(&count2) {
```

Comment:
`get_latest_valid_seq` returns the number of log items from latest file? Why is it used to determine the synchronize direction?